exonum/src/storage/proof_map_index/proof.rs

```diff
@@ -44,538 +42,841 @@ impl Serialize for ProofPath {
+                    }
+                }
+
+                Ok(ProofPath::new(&bytes).prefix(len as u16))
+            }
+        }
+
+        deserializer.deserialize_str(ProofPathVisitor)
+    }
 }
 
-/// An enum that represents a branch node of the map proof.
-pub enum BranchProofNode<V> {
-    /// A branch of proof in which both children do not contain the requested key.
-    BranchKeyNotFound {
-        /// A hash of the left child.
-        left_hash: Hash,
-        /// A hash of the right child.
-        right_hash: Hash,
-        /// A key of the left child.
-        left_key: ProofPath,
-        /// A key of the right child.
-        right_key: ProofPath,
-    },
-    /// A branch of proof in which left child may contain the requested key.
-    LeftBranch {
-        /// A left child node.
-        left_node: Box<ProofNode<V>>,
-        /// A hash of the right child.
-        right_hash: Hash,
-        /// A key of the left child.
-        left_key: ProofPath,
-        /// A key of the right child.
-        right_key: ProofPath,
-    },
-    /// A branch of proof in which right child may contain the requested key.
-    RightBranch {
-        /// A hash of the left child.
-        left_hash: Hash,
-        /// A right child node.
-        right_node: Box<ProofNode<V>>,
-        /// A key of the left child.
-        left_key: ProofPath,
-        /// A key of the right child.
-        right_key: ProofPath,
+/// An error returned when a map proof is invalid.
+#[derive(Debug, Fail)]
+pub enum MapProofError {
+    /// Non-terminal node for a map consisting of a single node.
+    #[fail(display = "non-terminal node as a single key in proof")]
+    NonTerminalNode(ProofPath),
+
+    /// One path in the proof is a prefix of another path.
+    #[fail(display = "embedded paths in proof")]
+    EmbeddedPaths {
+        /// Prefix key.
+        prefix: ProofPath,
+        /// Key containing the prefix.
+        path: ProofPath,
     },
+
+    /// One path is mentioned several times in the proof.
+    #[fail(display = "duplicate path in proof")]
+    DuplicatePath(ProofPath),
+
+    /// Entries in the proof are not ordered by increasing path.
+    #[fail(display = "invalid path ordering")]
+    InvalidOrdering(ProofPath, ProofPath),
 }
 
-impl<V: StorageValue> MapProof<V> {
-    /// Returns Merkle root hash of the map proof.
-    pub fn merkle_root(&self) -> Hash {
-        use self::MapProof::*;
+// Used instead of `(ProofPath, Hash)` only for the purpose of clearer (de)serialization.
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+struct MapProofEntry {
+    path: ProofPath,
+    hash: Hash,
+}
+
+// Used instead of `(K, Option<V>)` only for the purpose of clearer (de)serialization.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+enum OptionalEntry<K, V> {
+    Missing { missing: K },
+    KV { key: K, value: V },
```

Comment:
It looks like huge over-engineering to me. Why not `struct OptionalEntry { K, Option<V> }`?

Reply:
The major reason is that if we ever support storing `Option<V>` as `StorageValue`, it would be impossible to tell `(key, None)` and `(key, Some(None))` in JSON serialization. I agree this concern is somewhat artificial, but nevertheless I'd like to split missing and existing entries, so they cannot be mixed up. Another way would be to avoid overt serialization of missing entries completely, but this makes proofs less "transferable" in a sense that it would become impossible to tell what exactly the proof proves.

Reply:
Thanks, I've been thinking about this today, now I'm convinced.